rust/cymbal/src/fingerprinting/v1.rs

```diff
@@ -0,0 +1,264 @@
+use crate::{
+    error::Error,
+    metric_consts::ERRORS,
+    types::{frames::Frame, Exception},
+};
+use reqwest::Url;
+use sha2::{Digest, Sha256};
+
+// Given resolved Frames vector and the original Exception, we can now generate a fingerprint for it
+pub fn generate_fingerprint(
+    exception: &Exception,
+    mut frames: Vec<Frame>,
+) -> Result<String, Error> {
+    let mut fingerprint = format!(
+        "{}-{}",
+        exception.exception_type, exception.exception_message
+    );
+
+    let has_resolved_frames: bool = frames.iter().any(|f| f.resolved);
+    if has_resolved_frames {
+        frames.retain(|f| f.resolved);
+    }
+
+    let has_in_app_frames: bool = frames.iter().any(|f| f.in_app);
+    if has_in_app_frames {
+        frames.retain(|f| f.in_app);
+    } else {
+        metrics::counter!(ERRORS, "cause" => "no_in_app_frames").increment(1);
+        frames = frames.into_iter().take(1).collect()
+    }
+
+    for frame in frames {
+        let source_fn = match Url::parse(&frame.source.unwrap_or("".to_string())) {
+            Ok(url) => url.path().to_string(),
+            Err(_) => "unknown".to_string(),
+        };
+
+        fingerprint.push('-');
+        fingerprint.push_str(&source_fn);
+        fingerprint.push(':');
```

Comment:
Curious why we chose this mix of delineators? I don't have a necessarily better suggestion but the resultant string isn't particularly readable (`undefined-/alpha/foo.js:bar-/bar.js:baz-unknown:xyz` in the test). Perhaps that doesn't matter if it's going to be hashed anyways